test/CircularityTest.ts

```diff
@@ -0,0 +1,72 @@
+import "mocha";
+import { Microgrammar } from "../src/Microgrammar";
+import { firstOf } from "../src/Ops";
+import { JAVA_IDENTIFIER } from "./integration/RealWorldTest2";
+
+import * as assert from "power-assert";
+import { fail } from "power-assert";
+import { Concat } from "../src/matchers/Concat";
+import { isPatternMatch } from "../src/PatternMatch";
+
+describe("Circularity", () => {
+
+    const simpleType = {
+        name: JAVA_IDENTIFIER,
+    };
+
+    const genericType = {
+        name: JAVA_IDENTIFIER,
+        _l: "<",
+        param: null,    // This will be set later
+        _r: ">",
+        $lazy: true,
+    };
+
+    const genericType2 = Concat.of(genericType);
+
+    const type = firstOf(genericType2, simpleType);
+    // Set references
+    genericType.param = type;
+    genericType2._init();
```

Comment:
This would be the simplest programming model but it's a threading nightmare and will also be performance overhead